### ISSUE TITLE
SameDiff squaredNorm backprop fails

### ISSUE THREAD

**Issue by rnett** [deeplearning4j/deeplearning4j#7401]
#### Issue Description

SameDiff `squaredNorm` backprop fails under certain circumstances.  Haven't been able to reproduce in isolation.

Specifically, the input array looks like `[64,8,8]`, and the call looks like `squaredNorm(x, true, 2)`.

Stacktrace:
```java
[main] ERROR org.nd4j.linalg.cpu.nativecpu.ops.NativeOpExecutioner - Failed to execute op reduce_sqnorm_bp. Attempted to execute with 2 inputs, 1 outputs, 1 targs and 1 iargs. Inputs: [(FLOAT,[64,8,8],c), (FLOAT,[64,8,1],c)]. Outputs: [(FLOAT,[64,8,8],c)]. tArgs: [1.0]. iArgs: [2] - Please see above message (printed out from c++) for a possible cause of error.
java.lang.RuntimeException: NDArray::operator*=: the shapes of this and other arrays are not suitable for broadcast operation !
```
(There isn't a c++ error message above)

#### Version Information

Please indicate relevant versions, including, if relevant:

* Deeplearning4j version: 1.0.0-SNAPSHOT
* platform information (OS, etc): Windows 10
* CPU Backend

**Comment by rnett** [comment]
To reproduce:
```java
@Test
public void test4() {
    SameDiff SD = SameDiff.create();
    SDVariable in = SD.zero("test", 64, 8, 8);
    SDVariable squaredNorm = SD.squaredNorm("testOut", in, true, 2);

    SDVariable scale = SD.math.sqrt(squaredNorm.plus(1e-5));

    SDVariable out = in.times(scale).div("out", scale.times(squaredNorm.plus(1)));

    System.out.println(Arrays.toString(out.eval().shape()));

    SD.execBackwards(null, Lists.asList("test", new String[0]));
    SD.grad("test").eval(); // <-- fails here
}
```

Stacktrace looks like:
```java
[64, 8, 8]
Can't assign new value to the array: this shape [64, 8, 8]; other shape: [64, 8, 1]
[main] ERROR org.nd4j.linalg.cpu.nativecpu.ops.NativeOpExecutioner - Failed to execute op divide_bp. Attempted to execute with 3 inputs, 2 outputs, 0 targs and 0 iargs. Inputs: [(FLOAT,[64,8,8],c), (FLOAT,[64,8,1],c), (FLOAT,[],c)]. Outputs: [(FLOAT,[64,8,8],c), (FLOAT,[64,8,1],c)]. tArgs: -. iArgs: - - Please see above message (printed out from c++) for a possible cause of error.

java.lang.RuntimeException: Lengths of arrays are mismatched

	at org.nd4j.nativeblas.Nd4jCpu$NativeOps.execCustomOp(Native Method)
	at org.nd4j.linalg.cpu.nativecpu.ops.NativeOpExecutioner.exec(NativeOpExecutioner.java:1679)
	at org.nd4j.autodiff.samediff.internal.InferenceSession.getOutputs(InferenceSession.java:449)
	at org.nd4j.autodiff.samediff.internal.InferenceSession.getOutputs(InferenceSession.java:37)
	at org.nd4j.autodiff.samediff.internal.AbstractSession.output(AbstractSession.java:250)
	at org.nd4j.autodiff.samediff.SameDiff.exec(SameDiff.java:3357)
	at org.nd4j.autodiff.samediff.SDVariable.eval(SDVariable.java:1773)
	at layers.IssueTest.test4(IssueTest.java:27)
```

**Comment by AlexDBlack** [comment]
Make sure you pull the latest snapshots (or master, if you are building locally)
There were some broadcast fixes yesterday for issues like this - it might make been fixed already.

**Comment by rnett** [comment]
Yeah, I'm up to date, assuming the fixes have been pushed to the snapshot build.

**Comment by AlexDBlack** [comment]
OK, so there's actually 2 separate problems here. (Both present on moster)

First: reduce_sqnorm_bp is definitely broken for along dimension case - cc @shyrma 
```
CustomOp op = DynamicCustomOp.builder(s + "_bp")
                        .addInputs(Nd4j.create(DataType.FLOAT, 8, 6, 4), Nd4j.create(DataType.FLOAT, 8, 6, 1))
                        .addOutputs(Nd4j.create(DataType.FLOAT, 8, 6, 4))
                        .addBooleanArguments(true)  //Keep dim = true
                        .addIntegerArguments(2) //Dimension
                        .build();
                Nd4j.exec(op);
```
```
java.lang.RuntimeException: NDArray::operator*=: the shapes of this and other arrays are not suitable for broadcast operation !
```

Second: the problem stems from how SameDiff currently handles loss functions. Specifically, it is assumed that the final output of the network represents the loss function value - i.e., it should be a scalar. In this particular network, the output is a rank 3 array.

I'm not currently happy with this design... clearly, we need a loss function which must be a scalar, but the whole "last output must be scalar and is interpreted as the loss value" output thing should be improved. At the very least, we should provide some sort of check/warning

Anyway, if you change squared norm to norm2 (to work around first issue) and add an ```out = out.mean();``` to work around the second issue, the provided code runs fine.

**Comment by shugeo** [comment]
NDArray::operator *= cause an exception.

**Comment by shugeo** [comment]
Ok, fixed, you can to test and close it when not reproducible.

**Comment by AlexDBlack** [comment]
Confirmed passing now - thanks @shugeo
